agent/config/ratelimited_file_watcher.go

```diff
@@ -0,0 +1,85 @@
+package config
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+type rateLimitedFileWatcher struct {
+	watcher          Watcher
+	eventCh          chan *FileWatcherEvent
+	coalesceInterval time.Duration
+}
+
+func (r rateLimitedFileWatcher) Start(ctx context.Context) {
+	r.watcher.Start(ctx)
+	r.coalesceTimer(r.watcher.EventsCh(), r.coalesceInterval)
+}
+
+func (r rateLimitedFileWatcher) Stop() error {
+	return r.watcher.Stop()
+}
+
+func (r rateLimitedFileWatcher) Add(filename string) error {
+	return r.watcher.Add(filename)
+}
+
+func (r rateLimitedFileWatcher) Remove(filename string) {
+	r.watcher.Remove(filename)
+}
+
+func (r rateLimitedFileWatcher) Replace(oldFile, newFile string) error {
+	return r.watcher.Replace(oldFile, newFile)
+}
+
+func (r rateLimitedFileWatcher) EventsCh() chan *FileWatcherEvent {
+	return r.eventCh
+}
+
+func NewRateLimitedFileWatcher(configFiles []string, logger hclog.Logger, coalesceInterval time.Duration) (Watcher, error) {
+
+	watcher, err := NewFileWatcher(configFiles, logger)
+	if err != nil {
+		return nil, err
+	}
+	return rateLimitedFileWatcher{
+		watcher:          watcher,
+		coalesceInterval: coalesceInterval,
+		eventCh:          make(chan *FileWatcherEvent),
+	}, nil
+}
+
+func (r rateLimitedFileWatcher) coalesceTimer(inputCh chan *FileWatcherEvent, coalesceDuration time.Duration) {
+	var coalesceTimer *time.Timer = nil
+	sendCh := make(chan struct{})
+	FileWatcherEvents := make([]string, 0)
```

Comment:
```suggestion
	var fileWatcherEvents []string
```
Combo suggestion:

1. using a capitalized variable name in a method body is strange because normally that is how you indicate exported fields
2. Usually if you don't have a good sense for how big a slice should be, we usually start them out as the `nil` slice and let them grow via `append`, rather than explicitly allocating a zero-length slice which immediately will have to be reallocated on first append anyway.